docs/settings.rst
Settings
========

You can use the following settings to configure juntagrico

Contact Information
-------------------

ORGANISATION_NAME
^^^^^^^^^^^^^^^^^
  The short name of your organisation

  Type: String

  default value

  .. code-block:: python

    "Juntagrico"

ORGANISATION_NAME_CONFIG
^^^^^^^^^^^^^^^^^^^^^^^^
  Additional information to enrich the organisation name with the type of the organisation and its corresponding article

  Type: Dictionary

  default value

  .. code-block:: python

    {"type" : "",
        "gender" : ""}

ORGANISATION_LONG_NAME
^^^^^^^^^^^^^^^^^^^^^^
  The long version of your organisation, if you have one otherwise also use the short one
  
  Type: String

  default value

  .. code-block:: python

    "Juntagrico the best thing in the world"

ORGANISATION_ADDRESS
^^^^^^^^^^^^^^^^^^^^
  The address of your organisation
  
  Type: dict with the keys
  - name
  - street
  - number
  - zip
  - city
  - extra

  default value

  .. code-block:: python

    {"name":"Juntagrico", 
        "street" : "Fakestreet",
        "number" : "123",
        "zip" : "12456",
        "city" : "Springfield",
        "extra" : ""}

ORGANISATION_PHONE
^^^^^^^^^^^^^^^^^^
  The phone number for your organisation

  Type: string

  default value

  .. code-block:: python

    ""

INFO_EMAIL
^^^^^^^^^^
  The general email of your organisation

  Type: String

  default value

  .. code-block:: python

    "info@juntagrico.juntagrico"

SERVER_URL
^^^^^^^^^^
  The base url of your organisation (not the one where you run juntagrico)

  Type: String

  default value

  .. code-block:: python

    "www.juntagrico.juntagrico"

Accounting
----------

ORGANISATION_BANK_CONNECTION
^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  the bank connection information of your organisation
  
  Type: dict with the keys
  - PC (postkonto)
  - BIC
  - IBAN
  - NAME

  default value

  .. code-block:: python

    {"PC" : "01-123-5",
        "IBAN" : "CH 00 12345 67890 12345 67890 10",
        "BIC" : "BIC12345XX",
        "NAME" : "Juntagrico Bank",}

CURRENCY
^^^^^^^^
  The default currency used within the system

  Type: String

  default value

  .. code-block:: python

    "CHF"


External Documents
------------------

BUSINESS_REGULATIONS
^^^^^^^^^^^^^^^^^^^^
  URL to your business regulations document
  
  Type: String

  default value

  .. code-block:: python

    ""

BYLAWS
^^^^^^
  URL to your bylaws document
  
  Type: String

  default value

  .. code-block:: python

    ""

FAQ_DOC
^^^^^^^
  URL to your FAQ document

  Type: String

  default value

  .. code-block:: python

    ""

EXTRA_SUB_INFO
^^^^^^^^^^^^^^
  If you use extra subscriptions this describes the URL to the document describing them

  Type: String

  default value

  .. code-block:: python

    ""

ACTIVITY_AREA_INFO
^^^^^^^^^^^^^^^^^^
  URL to your document describing your activity areas

  Type: String

  default value

  .. code-block:: python

    ""


Business Year
-------------

BUSINESS_YEAR_START
^^^^^^^^^^^^^^^^^^^
  Defining the start of the business year

  Type: dict with the keys
  - day
  - month

  default value

  .. code-block:: python

    {"day":1, "month":1}


BUSINESS_YEAR_CANCELATION_MONTH
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  The date until you can cancel your subscriptions

  Type: Integer

  default value

  .. code-block:: python

    12


Membership
----------

ENABLE_REGISTRATION
^^^^^^^^^^^^^^^^^^^
  Decides if new member can register

  Type: Boolean

  default value

  .. code-block:: python

    True

BASE_FEE
^^^^^^^^
  Yearly fee for members without a subscription

  Type: String

  default value

  .. code-block:: python

    ""

MEMBERSHIP_END_MONTH
^^^^^^^^^^^^^^^^^^^^
  The month at which end the members can leave the organisation

  Type: Integer

  default value

  .. code-block:: python

    6

MEMBERSHIP_END_NOTICE_PERIOD
^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  The notice period in months a member needs to account for when cancelling the membership

  Type: Integer

  default value

  .. code-block:: python

    0

Shares
------

ENABLE_SHARES
^^^^^^^^^^^^^
  Enable all share related funtionality

  Type: String

  default value

  .. code-block:: python

    True

SHARE_PRICE
^^^^^^^^^^^
  Price of one share

  Type: String

  default value

  .. code-block:: python

    "250"


Jobs
----

ASSIGNMENT_UNIT
^^^^^^^^^^^^^^^
  The mode how assignments are counted: Valid values are EMTITY and HOURS. ENTITY the assignments are counted by occurrence, Hours the value of the assignments are counted by the actual time the user spent on a job.

  Type: String

  default value

  .. code-block:: python

    "ENTITY"

PROMOTED_JOB_TYPES
^^^^^^^^^^^^^^^^^^
  Types of jobs which should apear on start page

  Type: List of Strings

  default value

  .. code-block:: python

    []

PROMOTED_JOBS_AMOUNT
^^^^^^^^^^^^^^^^^^^^
  Amount of jobs which should be promoted on the start page

  Type: Integer

  default value

  .. code-block:: python

    2

Depot
-----

DEPOT_LIST_GENERATION_DAYS
^^^^^^^^^^^^^^^^^^^^^^^^^^
  Days on which the delivery list can be generated

  Type: List of Integers representing days of the week, where Monday is 0 and Sunday is 6.

  default value

  .. code-block:: python

    [0,1,2,3,4,5,6]



DEFAULT_DEPOTLIST_GENERATORS
^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  Generators used to generate the depot list. Generators need the method signature ``generator_name(*args, **options)``

  Type: List of Strings which define the different generators to be invoked

  default value

  .. code-block:: python

    ['juntagrico.util.depot_list.default_depot_list_generation']


Appearance
----------

VOCABULARY
^^^^^^^^^^
  Vocabulary dictionary for organisation specific words. _pl indicates the plural of a word. the member key describes the custom name you give your members. the member_type key describes what you call your member in accordance to your oganisation form.

  Type: Dictionary

  default value

  .. code-block:: python

    {
        'member': 'Mitglied',
        'member_pl' : 'Mitglieder',
        'assignment' : 'Arbeitseinsatz',
        'assignment_pl' : 'Arbeitseinsätze',
        'share' : 'Anteilschein',
        'share_pl' : 'Anteilscheine',
        'subscription' : 'Abo',
        'subscription_pl' : 'Abos',
        'co_member' : 'Mitabonnent',
        'co_member_pl' : 'Mitabonnenten',
        'price' : 'Betriebsbeitrag',
        'member_type' : 'Mitglied',
        'member_type_pl' : 'Mitglieder',
        'depot' : 'Depot',
        'depot_pl' : 'Depots',
        'package': 'Tasche',
    }
    
SUB_OVERVIEW_FORMAT
^^^^^^^^^^^^^^^^^^^
  Template and delimiter for formatting the subscription overview.

  default value

  .. code-block:: python

    {'delimiter': '|',
     'format': '{product}:{size}:{type}={amount}'
    }

STYLES
^^^^^^
  Define styles to be included on all pages.
  If the template key is set, the specified template will be loaded in the header of the page.
  In the static key a list of css files can be defined to be included.
  If both keys are defined the template is included before the static css files.

  default value

  .. code-block:: python

    {
        'template': '',
        'static': []
    }

SCRIPTS
^^^^^^^
  Define scripts to be included on all pages.
  If the template key is set, the specified template will be loaded in the scripts part of the page.
  In the static key a list of javascript files can be defined to be included.
  If both keys are defined the template is included before the static javascript files.

  default value

  .. code-block:: python

    {
        'template': '',
        'static': []
    }

FAVICON
^^^^^^^
  If you want to use a custom favicon this specifies the path for your favicon

  Type: String

  default value

  .. code-block:: python

    "/static/juntagrico/img/favicon.ico"

IMAGES
^^^^^^
  Defining the different images for core and job assignments etc

  default value

  .. code-block:: python

    {'status_100': '/static/juntagrico/img/status_100.png',
        'status_75': '/static/juntagrico/img/status_75.png',
        'status_50': '/static/juntagrico/img/status_50.png',
        'status_25': '/static/juntagrico/img/status_25.png',
        'status_0': '/static/juntagrico/img/status_0.png',
        'single_full': '/static/juntagrico/img/single_full.png',
        'single_empty': '/static/juntagrico/img/single_empty.png',
        'single_core': '/static/juntagrico/img/single_core.png',
        'core': '/static/juntagrico/img/core.png'}

BOOTSTRAP
^^^^^^^^^
  If you want to use a customized version of bootstrap this specifies the corresponding path for it

  Type: String

  default value

  .. code-block:: python

    "/static/juntagrico/external/bootstrap-3.3.1/css/bootstrap.min.css"


Email
-----

EMAILS
^^^^^^
  Defining the different email templates

  default value

  .. code-block:: python

    {
        'welcome': 'mails/welcome_mail.txt',
        'co_welcome': 'mails/welcome_added_mail.txt',
        'co_added': 'mails/added_mail.txt',
        'password': 'mails/password_reset_mail.txt',
        'j_reminder': 'mails/job_reminder_mail.txt',
        'j_canceled': 'mails/job_canceled_mail.txt',
        'confirm': 'mails/confirm.txt',
        'j_changed': 'mails/job_time_changed_mail.txt',
        'j_signup': 'mails/job_signup_mail.txt',
        'd_changed': 'mails/depot_changed_mail.txt',
        's_created': 'mails/share_created_mail.txt',
        'n_sub': 'mails/new_subscription.txt',
        's_canceled': 'mails/subscription_canceled_mail.txt',
        'm_canceled': 'mails/membership_canceled_mail.txt',
    }

MAIL_TEMPLATE
^^^^^^^^^^^^^
  Path to your custom html email template if you want to overwrite the look and feel of the html emails

  Type: String

  default value

  .. code-block:: python

    "mails/email.html"

DEFAULT_MAILER
^^^^^^^^^^^^^^
  The code to send mails. for more info see the code specified in the default value
  The setting ``'juntagrico.util.mailer.batch.Mailer'`` uses a built in batch mailer,
  that sends the emails to the "bcc" recipients in separate emails.
  See ``BATCH_MAILER`` to configure it.

  default value

  .. code-block:: python

    'juntagrico.util.mailer.default.Mailer'


BATCH_MAILER
^^^^^^^^^^^^^^
  Configuration for the batch mailer. These are only effective, if
  DEFAULT_MAILER is set to ``'juntagrico.util.mailer.batch.Mailer'``.
  ``batch_size`` is the number of emails, that is sent in one batch.
  When set to 1, all emails are sent using "to" instead of "bcc".
  ``wait_time`` is the interval in which the batches are sent.

  default value

  .. code-block:: python

    {
        'batch_size': 39,
        'wait_time': 65
    }


FROM_FILTER
^^^^^^^^^^^
  Allows overriding the "from" field of outgoing emails. This can be used to prevent sending emails with a sender of different domain than the SMTP server, which triggers most spam filters.
  The setting consists of a regular expression and a default replacement. If the regular expression does NOT match the default replacement is used as "from", and the origonal "from" is set as "reply to"

  default value

  .. code-block:: python

    {
        'filter_expression': '.*',
        'replacement_from': ''
    }

WHITELIST_EMAILS
^^^^^^^^^^^^^^^^
  List of regular expression to determine which email addresses should receive emails while DEBUG mode is enabled

  default value

  .. code-block:: python

    []


MAILER_RICHTEXT_OPTIONS
^^^^^^^^^^^^^^^^^^^^^^^
  Configuration overrides of the tinyMCE editor of the mailer view.
  See default config in ``static/juntagrico/js/initMailer.js``.

  default value:

  .. code-block:: python

    {}


GDPR
----

GDPR_INFO
^^^^^^^^^
  URL to your gdpr document

  Type: String

  default value

  .. code-block:: python

    ""

COOKIE_CONSENT
^^^^^^^^^^^^^^
  The text, confirm text, link text and url of the cookie consent

  default value

  .. code-block:: python

    {'text': _('{} verwendet folgende Cookies: session, csfr, cookieconsent.').format(Site.objects.get_current().name),
     'confirm_text': _('einverstanden'),
     'link_text': _('Hier findest du mehr zum Thema'),
     'url': '/my/cookies'
    }

Demo Settings
-------------

DEMO_USER
^^^^^^^^^
  If you run a demo setup and want to display the login name on the login page
  
  Type: String

  default value

  .. code-block:: python

    ''

DEMO_PWD
^^^^^^^^
  If you run a demo setup and want to display the password on the login page

  default value

  .. code-block:: python

    ''


docs/release_notes.rst
Release Notes
=============

1.3.7
-----
Has no migrations

Fixes
^^^^^
* Add mail button for depot admins

1.3.6
-----
Has no migrations

Fixes
^^^^^
* Assignments overviews only consider current business year
* Fix in export subscription
* Fix that users with filter permission can actually see the filters

Features & Improvements
^^^^^^^^^^^^^^^^^^^^^^^
* Inconsistencies include missing primary member as error

1.3.5
-----
Has no migrations

Fixes
^^^^^
* Fix for MIME-Type picky browser
* Fix datatables dom

1.3.4
-----
Has no migrations

Fixes
^^^^^
* Fix incorrect display of email count in subscription filters
* Member filter for depots only displays members with active subscription

Features & Improvements
^^^^^^^^^^^^^^^^^^^^^^^
* assignment counts include now planned assignments
* extracted assignment overviews into separate management list
* introduced search builder in new assignment management list
* moved general filters in corresponding management section in admin menu
* filter for all members and for active members
* streamlined the admin menu


1.3.3
-----
Has no migrations

Fixes
^^^^^
* Fix in depot list amount overview
* rollback of xhtml2pdf version since it is not compatible with alpine linux

Features & Improvements
^^^^^^^^^^^^^^^^^^^^^^^
* Introduces subscription excel export



1.3.2
-----
Has no migrations

Fixes
^^^^^
* Fix subscription count and depot display in member filter list
* Fix saving subscriptions with limited permissions
* Fix error when adding existing co-members with active subscriptions

Features & Improvements
^^^^^^^^^^^^^^^^^^^^^^^
* WHITELIST_EMAILS supports now regular expressions and is now documented
* Deeper subscription inconsistency checks

1.3.1
-----
Has no migrations

Features & Improvements
^^^^^^^^^^^^^^^^^^^^^^^
* Developer Features:
   * Upgraded requirements


1.3.0
-----
Has  migrations

Features & Improvements
^^^^^^^^^^^^^^^^^^^^^^^
* Member Features:
   * Share management: Overview of shares & cancellation of individual shares
   * Subscription management: Option to leave a subscription
   * Show membership state on membership page
   * Set nickname for subscription as shown on depot list
   * More readable listing of subscription parts
   * Include location in job reminder email
* Admin Features:
   * History of subscription memberships & scheduling of future changes of subscription recipients
   * Shares now store the creation date
   * Filter members by permissions
   * Batch editing of dates in shares
   * Also copy unlimited places in job copying
   * Show content of future subscriptions in name
   * Check all timestamps for consistent order
* Improvements in depot list:
   * Layout optimization
   * Added vocabulary for "package" in depot list
* Developer Features:
   * Upgraded to django 3.1.x, therefore add 'django.template.context_processors.request' to the TEMPLATES setting under 'context_processors'
   * Added permissions for admin notifications on subscription part change
   * Removed job_id argument from contact-member view

Fixes
^^^^^
* Fixes in cancellation of extra subscriptions
* Hide unused products in subscription order form
* Fixed share count in subscription overview
* Fixed issue in job copying
* Fixed notification emails with no recipients
* Assert that active subscriptions have at least 1 active part
* Fix in cancellation of subscription parts
* Minor bugfixes and fixed typos

juntagrico/config.py
from typing import Any

from django.conf import settings
from django.contrib.sites.models import Site
from django.templatetags.static import static
from django.utils.translation import gettext as _


def _get_setting(setting_key, default: Any = ''):
    return lambda: getattr(settings, setting_key, default() if callable(default) else default)


def _get_setting_with_key(setting_key, default):
    def inner(key):
        if hasattr(settings, setting_key) and key in getattr(settings, setting_key):
            return getattr(settings, setting_key)[key]
        d = default[key]
        return d() if callable(d) else d

    return inner


def fallback_static(path):
    try:
        return static(path)
    except ValueError:
        return path


class Config:
    # organisation settings
    vocabulary = _get_setting_with_key(
        'VOCABULARY',
        {
            'member': _('Mitglied'),
            'member_pl': _('Mitglieder'),
            'assignment': _('Arbeitseinsatz'),
            'assignment_pl': _('Arbeitseinsätze'),
            'share': _('Anteilschein'),
            'share_pl': _('Anteilscheine'),
            'subscription': _('Abo'),
            'subscription_pl': _('Abos'),
            'co_member': _('MitabonnentIn'),
            'co_member_pl': _('MitabonnentInnen'),
            'price': _('Betriebsbeitrag'),
            'member_type': _('Mitglied'),
            'member_type_pl': _('Mitglieder'),
            'depot': _('Depot'),
            'depot_pl': _('Depots'),
            'package': _('Tasche'),
        }
    )
    organisation_name = _get_setting('ORGANISATION_NAME', 'Juntagrico')
    organisation_name_config = _get_setting('ORGANISATION_NAME_CONFIG', {'type': '', 'gender': ''})
    organisation_long_name = _get_setting('ORGANISATION_LONG_NAME', 'Juntagrico the best thing in the world')
    organisation_address = _get_setting(
        'ORGANISATION_ADDRESS',
        {
            'name': 'Juntagrico',
            'street': 'Fakestreet',
            'number': '123',
            'zip': '12456',
            'city': 'Springfield',
            'extra': ''
        }
    )
    organisation_phone = _get_setting('ORGANISATION_PHONE')
    organisation_bank_connection = _get_setting(
        'ORGANISATION_BANK_CONNECTION',
        {
            'PC': '01-123-5',
            'IBAN': 'CH 00 12345 67890 12345 67890 10',
            'BIC': 'BIC1234500',
            'NAME': 'Juntagrico Bank',
        }
    )
    share_price = _get_setting('SHARE_PRICE', '250')
    business_regulations = _get_setting('BUSINESS_REGULATIONS')
    bylaws = _get_setting('BYLAWS')
    gdpr_info = _get_setting('GDPR_INFO')
    faq_doc = _get_setting('FAQ_DOC')
    extra_sub_info = _get_setting('EXTRA_SUB_INFO')
    activity_area_info = _get_setting('ACTIVITY_AREA_INFO')
    enable_shares = _get_setting('ENABLE_SHARES', True)
    enable_registration = _get_setting('ENABLE_REGISTRATION', True)
    base_fee = _get_setting('BASE_FEE')
    currency = _get_setting('CURRENCY', 'CHF')
    assignment_unit = _get_setting('ASSIGNMENT_UNIT', 'ENTITY')
    promoted_job_types = _get_setting('PROMOTED_JOB_TYPES', [])
    promomted_jobs_amount = _get_setting('PROMOTED_JOBS_AMOUNT', 2)
    depot_list_generation_days = _get_setting('DEPOT_LIST_GENERATION_DAYS', [0, 1, 2, 3, 4, 5, 6])
    default_depot_list_generators = _get_setting('DEFAULT_DEPOTLIST_GENERATORS', ['juntagrico.util.depot_list.default_depot_list_generation'])
    business_year_start = _get_setting('BUSINESS_YEAR_START', {'day': 1, 'month': 1})
    business_year_cancelation_month = _get_setting('BUSINESS_YEAR_CANCELATION_MONTH', 12)
    membership_end_month = _get_setting('MEMBERSHIP_END_MONTH', 6)
    membership_end_notice_period = _get_setting('MEMBERSHIP_END_NOTICE_PERIOD', 0)
    cookie_consent = _get_setting_with_key(
        'COOKIE_CONSENT',
        {
            'text': lambda: _('{} verwendet folgende Cookies: session, csfr, cookieconsent.').format(
                Site.objects.get_current().name),
            'confirm_text': _('einverstanden'),
            'link_text': _('Hier findest du mehr zum Thema'),
            'url': '/my/cookies'
        }
    )
    sub_overview_format = _get_setting_with_key(
        'SUB_OVERVIEW_FORMAT',
        {
            'delimiter': '|',
            'format': '{product}:{size}:{type}={amount}',
            'part_format': '{size}'
        }
    )

    # url and email settings
    info_email = _get_setting('INFO_EMAIL', 'info@juntagrico.juntagrico')
    contacts = _get_setting_with_key(
        'CONTACTS',
        {
            'general': lambda: Config.info_email(),
            'for_members': lambda: Config.contacts('general'),
            'for_subscriptions': lambda: Config.contacts('general'),
            'for_shares': lambda: Config.contacts('general'),
            'technical': lambda: Config.contacts('general'),
        }
    )
    server_url = _get_setting('SERVER_URL', 'www.juntagrico.juntagrico')
    default_mailer = _get_setting('DEFAULT_MAILER', 'juntagrico.util.mailer.default.Mailer')
    batch_mailer = _get_setting_with_key(
        'BATCH_MAILER',
        {
            'batch_size': 39,
            'wait_time': 65
        })
    from_filter = _get_setting_with_key('FROM_FILTER',
                                        {
                                            'filter_expression': '.*',
                                            'replacement_from': ''
                                        })

    # template settings
    mail_template = _get_setting('MAIL_TEMPLATE', 'mails/email.html')
    emails = _get_setting_with_key(
        'EMAILS',
        {
            'welcome': 'mails/member/member_welcome.txt',
            'co_welcome': 'mails/member/co_member_welcome.txt',
            'co_added': 'mails/member/co_member_added.txt',
            'password': 'mails/member/password_reset.txt',
            'confirm': 'mails/member/email_confirm.txt',
            'j_reminder': 'mails/member/job_reminder.txt',
            'j_canceled': 'mails/member/job_canceled.txt',
            'j_changed': 'mails/member/job_time_changed.txt',
            'j_signup': 'mails/member/job_signup.txt',
            'd_changed': 'mails/member/depot_changed.txt',
            's_created': 'mails/member/share_created.txt',
            'm_left_subscription': 'mails/member/co_member_left_subscription.txt',
            'n_sub': 'mails/admin/subscription_created.txt',
            's_canceled': 'mails/admin/subscription_canceled.txt',
            'a_share_created': 'mails/admin/share_created.txt',
            'a_share_canceled': 'mails/admin/share_canceled.txt',
            'a_subpart_created': 'mails/admin/subpart_created.txt',
            'a_subpart_canceled': 'mails/admin/subpart_canceled.txt',
            'a_member_created': 'mails/admin/member_created.txt',
            'a_depot_list_generated': 'mails/admin/depot_list_generated.txt',
            'm_canceled': 'mails/admin/member_canceled.txt',
        }
    )
    favicon = _get_setting('FAVICON', fallback_static('juntagrico/img/favicon.ico'))
    bootstrap = _get_setting('BOOTSTRAP', fallback_static('juntagrico/external/bootstrap-4.6.2/css/bootstrap.min.css'))
    styles = _get_setting_with_key('STYLES', {'template': '', 'static': []})
    scripts = _get_setting_with_key('SCRIPTS', {'template': '', 'static': []})
    images = _get_setting_with_key(
        'IMAGES',
        {
            'status_100': fallback_static('juntagrico/img/status_100.png'),
            'status_75': fallback_static('juntagrico/img/status_75.png'),
            'status_50': fallback_static('juntagrico/img/status_50.png'),
            'status_25': fallback_static('juntagrico/img/status_25.png'),
            'status_0': fallback_static('juntagrico/img/status_0.png'),
            'single_full': fallback_static('juntagrico/img/single_full.png'),
            'single_empty': fallback_static('juntagrico/img/single_empty.png'),
            'single_core': fallback_static('juntagrico/img/single_core.png'),
            'core': fallback_static('juntagrico/img/core.png')
        }
    )
    mailer_richtext_options = _get_setting('MAILER_RICHTEXT_OPTIONS', {})

    # demo settings
    demouser = _get_setting('DEMO_USER')
    demopwd = _get_setting('DEMO_PWD')
